### PR TITLE
[IE][VPU]: Fixes BinaryEltwise DTS on empty input

### PR DESCRIPTION
**Pull request by gladilov-gleb**
# Description

Makes DTS for BinaryEltwise produce empty output tensor in case if at least one input is empty. As a criteria for empty tensor ReduceMin is used (assuming all shape's values are non-negative).

Tests are changed accordingly. Trying to add new test case on inference with empty input, reference version failed, so those tests are left unchanged.

# Task

#-46858

### CODE REVIEW

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp [hunk 1]
```diff
@@ -48,7 +48,25 @@ void processBinaryEltwise(std::shared_ptr<ngraph::Node> eltwise, size_t lhsIndex
...
+        eltwise->get_friendly_name(), eltwise->get_type_info(), rhsInputShape);
+
+    const auto& lhsMin = std::make_shared<ngraph::opset5::ReduceMin>(lhsInput, ngraph::opset5::Constant::create(shapeElementType, {}, {0}));
+    const auto& rhsMin = std::make_shared<ngraph::opset5::ReduceMin>(rhsInput, ngraph::opset5::Constant::create(shapeElementType, {}, {0}));
+    const auto maxRank = std::max(ngraph::shape_size(lhsInputShape.to_shape()), ngraph::shape_size(lhsInputShape.to_shape()));
+    const auto& zeros = std::vector<std::int64_t>(maxRank, 0);
+    const auto& shape = std::make_shared<ngraph::opset5::Select>(
+        std::make_shared<ngraph::opset5::Convert>(std::make_shared<ngraph::opset5::Minimum>(lhsMin, rhsMin), ngraph::element::boolean),
+        std::make_shared<ngraph::opset3::Maximum>(lhsInput, rhsInput),
+        ngraph::opset5::Constant::create(shapeElementType, {static_cast<std::size_t>(maxRank)}, zeros));
```
Comment (gladilov-gleb) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:69]: @slyalin @lazarevevgeny 

What should we do if we're trying to execute Add on inputs with shapes {0} and {1}?

Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:69]: Discussed offline. {1} is broadcasted to {0} so solution meets specification criteria.

**Comment by gladilov-gleb** [comment]
@andrejsokolov @AndrewBakalinIntel ping

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp [hunk 2]
```diff
@@ -48,7 +48,25 @@ void processBinaryEltwise(std::shared_ptr<ngraph::Node> eltwise, size_t lhsIndex
...
+
+    VPU_THROW_UNLESS(lhsInputShape.is_static() && lhsInputShape.rank().get_length() == 1,
+        "DynamicToStaticShape transformation for {} of type {} expects lhs input shape to be static 1D vector, actual is {}",
+        eltwise->get_friendly_name(), eltwise->get_type_info(), lhsInputShape);
+
+    VPU_THROW_UNLESS(rhsInputShape.is_static() && rhsInputShape.rank().get_length() == 1,
+        "DynamicToStaticShape transformation for {} of type {} expects rhs input shape to be static 1D vector, actual is {}",
+        eltwise->get_friendly_name(), eltwise->get_type_info(), rhsInputShape);
+
+    const auto& lhsMin = std::make_shared<ngraph::opset5::ReduceMin>(lhsInput, ngraph::opset5::Constant::create(shapeElementType, {}, {0}));
```
Comment (AndrewBakalinIntel) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:62]: Why do you need this `ReduceMin`? It looks like we will get the output shape full of zeros in case if we have at least one zero in one of input shapes. Is it expected?

**Review by AndrewBakalinIntel — COMMENTED** [review]
Could you please also add a couple of tests on inference?

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp (hunk 2 above)
Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:62]: If we have at least one zero in any of input shapes it means we have an empty input, right?
If we have an empty input it means second input must broadcast to empty shape (discussed with @slyalin and @lazarevevgeny)

Regarding shape full of zeros: it's a good question which shape exactly must output have in empty input scenario. In the sake of simplicity I preserver only rank and fill all dimensions with zero. If anyone has some concerns about that I'm happy to discuss rationale and possible implementation.

Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:62]: > Why do you need this ReduceMin?

I wanted to use `ReduceProd` at first, but this operation is not support in Myriad-X plugin, so used `ReduceMin` assuming dimensions are non-negative.

**Comment by gladilov-gleb** [comment]
> Could you please also add a couple of tests on inference?

I tried to add some inference tests with empty input, but reference implementation failed for some reason on those cases (discussed offline with @itikhono).

@mikhail-treskin do you know anything about it? Should I raise a ticket on this?

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp (hunk 2 above)
Comment (AndrewBakalinIntel) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:62]: I think we need to do this based on the broadcast specification.
In case we have something like [100, 1, 0, 50] and [1], I expect to see [100, 1, 0, 50] as the output shape, because if some of the dimensions is broadcasting to zero, it doesn't mean that the rule changes for other dimensions.
Also some of these output dimensions may be extracted via `ShapeOf` afterwards, so it may affect some cases on inference.

**Comment by gladilov-gleb** [comment]
> > Could you please also add a couple of tests on inference?
> 
> I tried to add some inference tests with empty input, but reference implementation failed for some reason on those cases (discussed offline with @itikhono).
> 
> @mikhail-treskin do you know anything about it? Should I raise a ticket on this?

@mikhail-treskin ping

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp (hunk 2 above)
Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp]: Please, take a look at my solution once more

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp [hunk 3]
```diff
@@ -48,9 +49,51 @@ void processBinaryEltwise(std::shared_ptr<ngraph::Node> eltwise, size_t lhsIndex
...
+        "DynamicToStaticShape transformation for {} of type {} expects rhs input shape to be static 1D vector, actual is {}",
+        eltwise->get_friendly_name(), eltwise->get_type_info(), rhsInputShape);
+
+    std::shared_ptr<ngraph::Node> maxShape = std::make_shared<ngraph::opset3::Maximum>(lhsInput, rhsInput);
+    const auto& maxShapePartialShape = maxShape->output(0).get_partial_shape();
+    VPU_THROW_UNLESS(maxShapePartialShape.is_static(),
+        "DynamicToStaticShape transformation for {} of type {} expects output shape to have static rank, actual is dynamic");
+    const auto& maxShapeRank = ngraph::shape_size(maxShapePartialShape.to_shape());
+
+    const auto& updateOutputShapeOnZerosFrom = [&maxShape, &maxShapeRank, &shapeElementType](const ngraph::Output<ngraph::Node>& input) {
```
Comment (AndrewBakalinIntel) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:69]: Just to clarify, if we assume that we don't have negative shape values (as I can see from the description), should this code also work?
```
const auto& minimum = std::make_shared<ngraph::opset6::Minimum>(lhsInput, rhsInput);
outShape = std::make_shared<ngraph::opset6::Select>(
    std::make_shared<ngraph::opset6::Convert>(minimum, ngraph::element::boolean),
    maxShape,
    zeros);
```

Or there are some corner cases which I did not consider?

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp [hunk 4]
```diff
@@ -48,9 +49,51 @@ void processBinaryEltwise(std::shared_ptr<ngraph::Node> eltwise, size_t lhsIndex
...
+        const auto& shapeValue = input.get_partial_shape();
+        const auto& rank = ngraph::shape_size(shapeValue.to_shape());
+
+        const auto& zeros = ngraph::opset5::Constant::create(shapeElementType, {rank}, std::vector<std::int64_t>(rank, 0));
```
Comment (AndrewBakalinIntel) [inline inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:73]: I think it's better to use the latest opset if possible which is `opset6`

**Review by AndrewBakalinIntel — COMMENTED** [review]
Please update copyrights as well.

**Comment by AndrewBakalinIntel** [comment]
> > Could you please also add a couple of tests on inference?
> 
> I tried to add some inference tests with empty input, but reference implementation failed for some reason on those cases (discussed offline with @itikhono).
> 
> @mikhail-treskin do you know anything about it? Should I raise a ticket on this?

Do you mean static cases or dynamic? Or they both failed?

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp [hunk 5]
```diff
@@ -48,9 +49,51 @@ void processBinaryEltwise(std::shared_ptr<ngraph::Node> eltwise, size_t lhsIndex
...
+        const auto& isZero = std::make_shared<ngraph::opset5::Equal>(input, zeros);
+        if (maxShapeRank == rank) {
+            maxShape = std::make_shared<ngraph::opset5::Select>(isZero, zeros, maxShape);
+        } else {
```
[the remark this answers is not in the document]
Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:77]: Excellent catch! I though `Maximum` operation does that for us

**Comment by gladilov-gleb** [comment]
> > > Could you please also add a couple of tests on inference?
> > 
> > 
> > I tried to add some inference tests with empty input, but reference implementation failed for some reason on those cases (discussed offline with @itikhono).
> > @mikhail-treskin do you know anything about it? Should I raise a ticket on this?
> 
> Do you mean static cases or dynamic? Or they both failed?

Dynamic

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp (hunk 3 above)
Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp:69]: Looks like this should work, but uses more operations:
Constant + Minimum + Select + Convert vs Constant + Equal + Select

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp]: Fixed

File: inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp (hunk 4 above)
Comment (gladilov-gleb) [reply inference-engine/src/vpu/common/src/ngraph/transformations/dynamic_to_static_shape_binary_elementwise.cpp]: Fixed

**Comment by gladilov-gleb** [comment]
@andrejsokolov please take a look

**Review by AndrewBakalinIntel — APPROVED** [review]

**Review by andrejsokolov — APPROVED** [review]
